content/blog/2019-11-13-null-pointer-guards.md

```diff
@@ -0,0 +1,220 @@
+`p` aliases `&a`. That is, the memory location that `p` point to and `&a`
+point to are the same. Thus, when we do `*p = nullptr`, we are also setting
+`a` to `nullptr`. This means that whenever we a store a value `%p` to `%deref_q`
+(the location pointed to by `%q`), we must change everything that aliases `%deref_q`.
+
+One strange thing about this implementation is how to represent `%deref_p`.
+Given `store i32* null, i32** %p, align 8`, we need to make sure that subsequent
+loads from `%p` are `PossiblyNull`. So, we keep a map `deref_map` from pointers like `%p` to
+an arbitrary new pointer. Therefore, when we see `%val = load i32*, i32** %p, align 8`,
+we set `lattice[val] = lattice[deref_map[p]]`, where `lattice[x]` tells us if
+`x` is `PossiblyNull` or `DefinitelyNonNull`.
+
+## Evaluation
+For correctness, I wrote some tests by hand to account for some of the
+programs shown above. For the most part, the correctness of adding the nullchecks
+was easy to check. However, the correctness of the NPA was trickier to determine,
+as shown by the programs that require an alias analysis.
+
+We also want to see how much of an impact these null checks have on the performance
+of programs. Unfortunately I wasn't able to get the PARSEC benchmark tests running.
+Instead, I found benchmarks from https://benchmarksgame-team.pages.debian.net/benchmarksgame/
+which has various benchmarks for different languages. I chose a small portion of
+these tests to run with and without my compiler passes.
+
+I chose to run:
+- [binary-trees](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/binarytrees-gpp-2.html)
+  - Ran with: `./binary-trees 13`
+- [fannkuch-redux](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gpp-5.html)
+  - Ran with: `./fannkucuh-redux 10`
+- [fasta](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fasta-gpp-1.html)
+  - Ran with: `./fasta 250000`
+- [mandelbrot](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/mandelbrot-gpp-5.html)
+  - Ran with: `./mandelbrot 2000`
+- [n-body](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-gpp-1.html)
+  - Ran with: `./nbody 5000000`
+
+Here we simply show the mean time taken to run these programs and the standard
+deviation. There are three columns, representing the three different ways we
+compiled the program.
+- *Baseline* is just compiling using Clang without our null check passes.
+Compiled with `/usr/local/opt/llvm/bin/clang++ -O2 <benchmark>.cpp -o <benchmark>`, e.g.,
+`/usr/local/opt/llvm/bin/clang++ -O2 binary-trees.cpp -o binary-trees`
+- *No NPA* is compiling using the null check passes, but without taking into
+consideration which pointers are `PossiblyNull` or `DefinitelyNonNull`. That is,
+we do a null check on every pointer dereference.
+NOTE: Had to go into source code and remove the check for if pointers are `PossiblyNull`
```

Comment:
It's probably possible/nicer to phrase this as a complete sentence, without the "NOTE:" marker? Just say, "To get this to work, I had to…"